src/System.Private.CoreLib/shared/System/Index.cs

```diff
@@ -28,10 +28,14 @@ public override int GetHashCode()
             return _value;
         }
 
-        public override string ToString()
+        public override string ToString() => FromEnd ? ToStringFromEnd() : ((uint)Value).ToString();
+
+        private string ToStringFromEnd()
         {
-            string str = Value.ToString();
-            return FromEnd ? "^" + str : str;
+            Span<char> span = stackalloc char[11]; // 1 for ^ and 10 for longest possible uint value
+            ((uint)Value).TryFormat(span.Slice(1), out int charsWritten);
+            span[0] = '^';
+            return new string(span.Slice(0, charsWritten + 1));
         }
```

Comment:
As an alternative - modified version of `UInt32ToDecStr` without CountDigits, etc
```csharp
private static unsafe string UInt32ToDecStr(uint value)
{
    char* buf = stackalloc char[11];
    int i = 11;
    do
    {
        uint div = value / 10;
        buf[--i] = (char)('0' + value - (div * 10));
        value = div;
    }
    while (value != 0);
    buf[--i] = '^';
    return new string(buf, i, 11 - i);
}
```

Reply:
@stephentoub yeah, surely doesn't make any sense, I was actually curious if your technique with `stackalloc[11]` could be used in `UInt32ToDecStr`, but looks like GetDigitsCount+string.FastAllocate is better 🙂